### PR TITLE
Enhancement: Implement Search #10

### PR DESCRIPTION
**Pull request by ycechungAI**
### Implements #10 
- [X] - bash script to run docker on local v2
- [X] - CI template to include bash script v2

### Follow Up Issues
#69 - have the proper dependencies and builds on CI, enabling user groups for docker?  You need python 3.6.15 or newer with pyenv installed.

#70 - Integration - https://docsearch.algolia.com/docs/legacy/dropdown
       - Maybe look into https://docsearch.algolia.com/docs/record-extractor - not sure how this works? v3

I loaded my own API and data for my own tests. If someone needs to test they need to create an account on Algolia and use their own app_id and key.  In the release of this site this will change as well.

Need to implement the CI to do indexing for the search.  I tried on Windows and not successful building manually because you need python 3.6 and certain things fail from there.  It is recommended to use Ubuntu on CI.  I will try again on Linux platform.

Refer to Blog post link for steps:

### References:
https://docsearch.algolia.com/docs/legacy/run-your-own/
https://discourse.algolia.com/t/how-to-fix-docsearch-docker-build-errors/6496/2

### CODE REVIEW

**Comment by ycechungAI** [comment]
image
i made a bash script called scrape.sh and here is a sample run

**Comment by ycechungAI** [comment]
the CI script uses Ubuntu and calls the bash script =)

File: package.json [hunk 1]
```diff
@@ -15,17 +15,18 @@
...
     "@mdx-js/react": "^1.6.21",
     "@svgr/webpack": "^5.5.0",
     "clsx": "^1.1.1",
     "file-loader": "^6.2.0",
     "prism-react-renderer": "^1.2.1",
     "react": "^17.0.1",
     "react-dom": "^17.0.1",
-    "url-loader": "^4.1.1"
+    "url-loader": "^4.1.1",
+    "yarn": "^1.22.17"
```
Comment (humphd) [inline package.json:29]: Why is this here?

File: package.json [hunk 2]
```diff
@@ -15,17 +15,18 @@
...
+    "@docusaurus/preset-classic": "^2.0.0-beta.9",
     "@mdx-js/react": "^1.6.21",
     "@svgr/webpack": "^5.5.0",
     "clsx": "^1.1.1",
     "file-loader": "^6.2.0",
     "prism-react-renderer": "^1.2.1",
     "react": "^17.0.1",
     "react-dom": "^17.0.1",
-    "url-loader": "^4.1.1"
+    "url-loader": "^4.1.1",
```
Comment (humphd) [inline package.json:28]: This isn't being used

File: scrape.sh [hunk 3]
```diff
@@ -0,0 +1,8 @@
+if [ -f .env ]; then
```
Comment (humphd) [inline scrape.sh:1]: Please add comments to this file.  Specifically, explain what each part is doing, and include links to upstream docs that help give more context.

File: .github/workflows/algolia-scraper.yml [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+  workflow_dispatch:
+jobs:
+  scrape:
+    runs-on: ubuntu-latest
+    steps:
+      - name: check out code 🛎
+        uses: actions/checkout@v2
+      - name: scrape the site 🧽
+        env:
+          APPLICATION_ID: ${{ secrets.APPLICATION_ID }}
```
Comment (humphd) [inline .github/workflows/algolia-scraper.yml:17]: Can you explain what will need to happen here?  We obviously need to add these keys to the repo.  What are they, how do we find them, etc?

**Review by humphd — CHANGES_REQUESTED** [review]

File: .github/workflows/algolia-scraper.yml (hunk 4 above)
Comment (ycechungAI) [reply .github/workflows/algolia-scraper.yml:17]: 1. you would need to sign up on www.aligolia.com and create a project for the purpose of creating index
2. add an index_name called "docusaurus" for instance
3. you add API_key and give rights for ACLs search, addObject, deleteObject, editSearch 
4. site_url is the deployed site, but siteindex needs to match to crawl it
5. if u look at the pull request ff0a9a5 i have a sample .env file with the variables set.  I removed it from the next PR and replaced it with process.env. - before it was hardcoded.  Ideally you want this public so u can dev but I think each dev should just create his own account because its free and owner of repo just create secrets for it then it would work

File: scrape.sh (hunk 3 above)
Comment (ycechungAI) [reply scrape.sh:1]: gottcha, basically if there is no .env file it will pull it from the secrets set on github
But if you want to actually build the index you create your own if the site is not deployed yet.

Comment (ycechungAI) [reply scrape.sh:1]: image

File: .github/workflows/algolia-scraper.yml (hunk 4 above)
Comment (humphd) [reply .github/workflows/algolia-scraper.yml:17]: Please put this into a section on `Search` in the `CONTRIBUTING.md` file, so we don't lose track of it.  There's no way we'll remember all this.

Comment (ycechungAI) [reply .github/workflows/algolia-scraper.yml:17]: i added comments with sample .env so u can see its approximate values fc91f68 so someone can easily follow up

**Comment by ycechungAI** [comment]
Im not sure how to comment in CI but here are the steps to algolia-scraper.yml
1. push: only run on push events
2. check_suite: only run on check suite events
3. workflow_dispatch: only run on workflow dispatch events
4. runs-on: specifies the OS image to use for the job
5. steps: specifies the commands to run
6. env: specifies the environment variables to use for the job
7. uses: specifies the action to use for the job
8. run: specifies the command to run in this case is running the scraper.sh

**Comment by humphd** [comment]
> Im not sure how to comment in CI but here are the steps to algolia-scraper.yml

See https://www.tutorialspoint.com/yaml/yaml_comments.htm

**Review by ycechungAI — COMMENTED** [review]
seen changes needed

File: .github/workflows/algolia-scraper.yml (hunk 4 above)
Comment (ycechungAI) [reply .github/workflows/algolia-scraper.yml]: > Please put this into a section on `Search` in the `CONTRIBUTING.md` file, so we don't lose track of it. There's no way we'll remember all this.

Added to CONTRIBUTING.md

**Comment by ycechungAI** [comment]
For now ill deal with the first two bullets as this is a growing issue. I leave the last three as road map for others to do.

**Comment by ycechungAI** [comment]
Filed followed up issues and moved the image here so its not blocking the top
image

**Review by ycechungAI — COMMENTED** [review]
done reading review

**Review by ycechungAI — COMMENTED** [review]
rereview changes

**Review by ycechungAI — COMMENTED** [review]
mark resolved

**Review by ycechungAI — COMMENTED** [review]
old review

File: .github/workflows/algolia-scraper.yml [hunk 5]
```diff
@@ -0,0 +1,25 @@
+name: algolia-scraper
+on:
+  push: # only run on push events
+    branches:
+      - main
+  check_suite: # only run check suite events
+    types: [completed]
+  workflow_dispatch: # only run on workflow dispatch events
```
Comment (humphd) [inline .github/workflows/algolia-scraper.yml:8]: What is this for?  I've never used it.  Why do we need it?

File: .github/workflows/algolia-scraper.yml [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+jobs:
+  scrape:
+    runs-on: ubuntu-latest # specifies the OS image to use for the job
+    steps: # specifies the commands to run
+      - name: check out code 🛎
```
Comment (humphd) [inline .github/workflows/algolia-scraper.yml:13]: Please follow the same style as we're using in https://github.com/Seneca-ICTOER/IPC144/pull/48/files#diff-259f2188de53828dcb003900d2a84cfd00a909870115a6e14ddc019e96255087R11, especially adding the `cache` for `yarn` so this goes faster.

File: .github/workflows/algolia-scraper.yml [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+        uses: actions/checkout@v2
+      - name: scrape the site 🧽
+        env: #- specifies the environment variables
+          APPLICATION_ID: ${{ secrets.APPLICATION_ID }} #generated when you create a project on algolia website after you sign up
```
Comment (humphd) [inline .github/workflows/algolia-scraper.yml:17]: In general, prefer putting comments above vs. beside your code, as it lengthens the line length unnecessarily.

File: .github/workflows/algolia-scraper.yml [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+          WRITE_API_KEY: ${{ secrets.API_KEY }} #API key be sure to have the right permissions
+          SITE_URL: ${{ secrets.SITE_URL }} #site url
+          INDEX_NAME: ${{ secrets.INDEX_NAME }} #docusaurus or anything if you want ot change it
+          # run: create .env file and run bash script scrape.sh
+        run: |
+          touch .env
```
Comment (humphd) [inline .github/workflows/algolia-scraper.yml:23]: What's the purpose of having an empty `.env`?

File: CONTRIBUTING.md [hunk 9]
```diff
@@ -66,3 +66,26 @@ GIT_USER=<Your GitHub username> USE_SSH=true yarn deploy
 ```
 
 If you are using GitHub pages for hosting, this command is a convenient way to build the website and push to the `gh-pages` branch.
+
+### Search
+
+To build index for search
+
+```text
+1. you would need to sign up an account on https://www.algolia.com/users/sign_in
```
Comment (humphd) [inline CONTRIBUTING.md:75]: `sign up for an account on ...`

File: CONTRIBUTING.md [hunk 10]
```diff
@@ -66,3 +66,26 @@ GIT_USER=<Your GitHub username> USE_SSH=true yarn deploy
 ```
 
 If you are using GitHub pages for hosting, this command is a convenient way to build the website and push to the `gh-pages` branch.
+
+### Search
+
+To build index for search
```
Comment (humphd) [inline CONTRIBUTING.md:72]: "We use Algolia to build our index.  In order to build an index for search:"

File: CONTRIBUTING.md [hunk 11]
```diff
@@ -66,3 +66,26 @@ GIT_USER=<Your GitHub username> USE_SSH=true yarn deploy
...
+
+```text
+1. you would need to sign up an account on https://www.algolia.com/users/sign_in
+2. create a project for the purpose of creating index
```
Comment (humphd) [inline CONTRIBUTING.md:76]: `...of creating the index` (add the)

File: CONTRIBUTING.md [hunk 12]
```diff
@@ -66,3 +66,26 @@ GIT_USER=<Your GitHub username> USE_SSH=true yarn deploy
...
+3. add an INDEX_NAME called "docusaurus" for instance on same website
```
Comment (humphd) [inline CONTRIBUTING.md:77]: wrap INDEX_NAME and "docusaurus" in backticks

File: CONTRIBUTING.md [hunk 13]
```diff
@@ -66,3 +66,26 @@ GIT_USER=<Your GitHub username> USE_SSH=true yarn deploy
...
+4. You add API_key and give rights for ACLs search, addObject, deleteObject, editSearch for the key
```
Comment (humphd) [inline CONTRIBUTING.md:78]: Wrap all the technical bits of this in backticks.  Also, confirm that it's API_key vs. API_KEY.

File: CONTRIBUTING.md [hunk 14]
```diff
@@ -66,3 +66,26 @@ GIT_USER=<Your GitHub username> USE_SSH=true yarn deploy
...
+5. SITE_URL is the deployed site, but siteindex needs to match it in order crawl it
+6. If u look at the pull request ff0a9a5 i have a sample .env file with the variables set.I also have a an example of what values needed as secret in scrape.sh
```
Comment (humphd) [inline CONTRIBUTING.md:80]: Use proper spelling, punctuation, and formal language:

```
If you look at the pull request ff0a9a5, I have a sample .env file with the variables set. I also have an example of which secret values are needed in scrape.sh
```

Also, link to anything that you're talking about.

File: CONTRIBUTING.md [hunk 15]
```diff
@@ -66,3 +66,26 @@ GIT_USER=<Your GitHub username> USE_SSH=true yarn deploy
...
+```
+
+Important reference websites for next steps
+
+```text
+https://docsearch.algolia.com/docs/integrations
```
Comment (humphd) [inline CONTRIBUTING.md:86]: These aren't text (i.e., no backticks), they should be a bulleted list of URLs.

File: docusaurus.config.js [hunk 16]
```diff
@@ -3,42 +3,47 @@ const darkCodeTheme = require('prism-react-renderer/themes/vsDark');
 
 /** @type {import('@docusaurus/types').DocusaurusConfig} */
 module.exports = {
-  title: 'IPC144 - Course Notes',
-  tagline: 'C Programming',
-  url: 'https://cghub.ca',
-  baseUrl: '/',
+  title: "IPC144 - Course Notes",
```
Comment (humphd) [inline docusaurus.config.js:6]: Please only change the parts of this file that you need to, don't reformat everything.

File: scrape.sh [hunk 17]
```diff
@@ -0,0 +1,22 @@
+#if the .env file exists, then export the environment variables from it
+#.env file sample:
+#=============================================
+#APPLICATION_ID=6TZGXYS27N
+#API_KEY=51151f6285b830cea1d7e40675f8af8f
```
Comment (humphd) [inline scrape.sh:5]: Make sure these aren't real.  If they are, delete them now (you shouldn't leak keys into github)

**Review by humphd — CHANGES_REQUESTED** [review]

File: .github/workflows/algolia-scraper.yml (hunk 5 above)
Comment (ycechungAI) [reply .github/workflows/algolia-scraper.yml:8]: CI to run the index generation?

**Review by ycechungAI — COMMENTED** [review]
looked at review

File: .github/workflows/algolia-scraper.yml (hunk 6 above)
Comment (ycechungAI) [reply .github/workflows/algolia-scraper.yml:13]: ok

File: .github/workflows/algolia-scraper.yml (hunk 8 above)
Comment (ycechungAI) [reply .github/workflows/algolia-scraper.yml:31]: the bash script needs to use those variables

**Review by ycechungAI — COMMENTED** [review]
refixed some things

**Review by ycechungAI — COMMENTED** [review]
reviewed docusaurus.config.js

File: docusaurus.config.js (hunk 16 above)
Comment (ycechungAI) [reply docusaurus.config.js:6]: ok

File: CONTRIBUTING.md [hunk 18]
```diff
@@ -6,11 +6,11 @@ Thanks for your interest in helping us maintain our course notes! We are happy t
 
 These notes are built using [Docusaurus](https://docusaurus.io/), a modern static website generator. To use it, you'll need to have the following installed:
 
-- an [LTS version of node.js](https://nodejs.org/en/), which can be checked by running `node -v`.  Currently that's node.js 14.x, but confirm with the official site.
+- an [LTS version of node.js](https://nodejs.org/en/), which can be checked by running `node -v`. Currently that's node.js 14.x, but confirm with the official site.
```
Comment (humphd) [inline CONTRIBUTING.md:9]: LTS is 16

File: CONTRIBUTING.md [hunk 19]
```diff
@@ -66,3 +66,25 @@ GIT_USER=<Your GitHub username> USE_SSH=true yarn deploy
 ```
 
 If you are using GitHub pages for hosting, this command is a convenient way to build the website and push to the `gh-pages` branch.
+
+### Search
+
+We use Algolia to build our index. In order to build an index for search:
+
+```text
```
Comment (humphd) [inline CONTRIBUTING.md:74]: Get rid of this.  Just use a numbered list

File: config.json [hunk 20]
```diff
@@ -0,0 +1,15 @@
+{
```
Comment (humphd) [inline config.json:1]: Can we do a better job of naming this?  If it's for algolia, let's call it algolia-config.json

File: docusaurus.config.js (hunk 16 above)
Comment (humphd) [inline docusaurus.config.js:6]: Please don't touch every line in a file for style changes, it makes reviewing harder than it needs to be and messes with git history.

File: package.json [hunk 21]
```diff
@@ -17,17 +17,17 @@
...
+    "@docusaurus/preset-classic": "^2.0.0-beta.9",
     "@mdx-js/react": "^1.6.21",
     "@svgr/webpack": "^5.5.0",
     "clsx": "^1.1.1",
     "file-loader": "^6.2.0",
     "prism-react-renderer": "^1.2.1",
     "react": "^17.0.1",
     "react-dom": "^17.0.1",
-    "url-loader": "^4.1.1"
+    "yarn": "^1.22.17"
```
Comment (humphd) [inline package.json:30]: Yarn isn't needed in here

**Review by humphd — CHANGES_REQUESTED** [review]
I'd love to get this finished and landed by tomorrow.

File: docusaurus.config.js (hunk 16 above)
Comment (ycechungAI) [reply docusaurus.config.js:6]: my linter was has a rule set to use double quotes I believe which changed the file automatically, I will fix my settings

**Review by japneetsingh035 — APPROVED** [review]
Reviewed: Tutorial for Converting Table-like Diagrams into Images

File: .github/workflows/algolia-scraper.yml [hunk 22]
```diff
@@ -0,0 +1,33 @@
+name: algolia-scraper
+on:
+  push:
+    branches: [main]
+    # only run on workflow dispatch events
```
Comment (humphd) [inline .github/workflows/algolia-scraper.yml:5]: Remove this comment

File: .github/workflows/algolia-scraper.yml [hunk 23]
```diff
@@ -0,0 +1,33 @@
...
+      - uses: actions/checkout@v2
+      - uses: actions/setup-node@v
+      - name: scrape the site 🧽
+        with:
+          node-version: '16.x'
+          cache: 'yarn'
+      - uses: webfactory/ssh-agent@v0.5.3
+        with:
+          ssh-private-key: ${{ secrets.GH_PAGES_DEPLOY }}
+      - name: build index using scrapper
```
Comment (humphd) [inline .github/workflows/algolia-scraper.yml:20]: "using Algolia Scraper" (sp)

**Review by humphd — CHANGES_REQUESTED** [review]
A couple small things, plus please squash and rebase this and I think we can merge.

@CameronGray1210: we'll have to set this up and add secrets to the repo for it to work, which we can do later.

**Comment by humphd** [comment]
Waiting on squash/rebase.

**Review by ycechungAI — COMMENTED** [review]
done rebase

File: scrape.sh [hunk 24]
```diff
@@ -0,0 +1,34 @@
+<<<<<<< HEAD
```
Comment (humphd) [inline scrape.sh:1]: You've got merge conflict markers in here

File: scrape.sh [hunk 25]
```diff
@@ -0,0 +1,34 @@
...
+#if the .env file exists, then export the environment variables from it
+#.env file sample:
+#=============================================
+#APPLICATION_ID=<>
+#API_KEY=<>
+#SITE_URL=https://ipc144.sdds.ca
+#INDEX_NAME=docusaurus
+#=============================================
+if [ -f .env ]; then
+>>>>>>> fc91f68... *removed url-loader *added comments to code *fixed some parts
```
Comment (humphd) [inline scrape.sh:22]: And here

File: scrape.sh [hunk 26]
```diff
@@ -0,0 +1,34 @@
...
+#if the .env file exists, then export the environment variables from it
+#.env file sample:
+#=============================================
+#APPLICATION_ID=<>
+#API_KEY=<>
+#SITE_URL=https://ipc144.sdds.ca
+#INDEX_NAME=docusaurus
+#=============================================
+f [ -f .env ]; then
+=======
```
Comment (humphd) [inline scrape.sh:12]: And here

**Review by humphd — CHANGES_REQUESTED** [review]
Did you use `npm` or `yarn` to install your deps?  It seems to have rewritten the entire lock file with npm links.

Can you delete `yarn.lock` and re-run `yarn install`?

**Comment by ycechungAI** [comment]
> Did you use `npm` or `yarn` to install your deps? It seems to have rewritten the entire lock file with npm links.
> 
> Can you delete `yarn.lock` and re-run `yarn install`?

I rebuild using yarn but there are broken links in the current build when i did yarn build

I re-based my commit and pushed to issue-10

**Comment by humphd** [comment]
The failed build for links is #106.

File: .github/workflows/algolia-scraper.yml [hunk 27]
```diff
@@ -0,0 +1,31 @@
+name: algolia-scraper
```
Comment (humphd) [inline .github/workflows/algolia-scraper.yml:1]: This needs to get added to the other deployment.yml file that we just merged, so it happens as part of a deployment.

**Review by humphd — CHANGES_REQUESTED** [review]

File: .github/workflows/algolia-scraper.yml (hunk 27 above)
Comment (ycechungAI) [reply .github/workflows/algolia-scraper.yml:1]: so is there any changes for me? I don't think i have access to the deployment.yml at this time

Comment (humphd) [reply .github/workflows/algolia-scraper.yml:1]: Rebase, and pick up https://github.com/Seneca-ICTOER/IPC144/blob/main/.github/workflows/deployment.yaml

Comment (ycechungAI) [reply .github/workflows/algolia-scraper.yml:1]: ok

**Review by ycechungAI — COMMENTED** [review]
reviewed waiting for changes to deployment.yml

**Comment by humphd** [comment]
When you say you're 'waiting for changes to deployment.yml', what does that mean?  That work for the deployment stuff is merged.  You need to move your workflow jobs into the same file so it all happens together.

Also, you need to rebase to pick up the change to docusaurus.config.js that just got merged.

**Comment by ycechungAI** [comment]
> When you say you're 'waiting for changes to deployment.yml', what does that mean? That work for the deployment stuff is merged. You need to move your workflow jobs into the same file so it all happens together.
> 
> Also, you need to rebase to pick up the change to docusaurus.config.js that just got merged.

i get it

File: .github/workflows/algolia-scraper.yml [hunk 28]
```diff
@@ -0,0 +1,31 @@
...
+on:
```
Comment (humphd) [inline .github/workflows/algolia-scraper.yml:2]: You're closer, but this is still wrong.  We want to run a series of `jobs` whenever we push to the main branch.  One of those jobs is to build the site, another is to index it, another is to deploy it.  The order matters, so we need them to all go in a single workflow, and depend on each other.  Your job needs to go in the same development.yml file that we already started, and be one part of the deploy step.

File: .gitignore [hunk 29]
```diff
@@ -8,6 +8,7 @@
 .docusaurus
 .cache-loader
 
+
```
Comment (humphd) [inline .gitignore:11]: I'm not sure why you're changing this file.

File: docs/F-Refinements/portability.md [hunk 30]
```diff
@@ -1,94 +1,99 @@
 ---
```
Comment (humphd) [inline docs/F-Refinements/portability.md:1]: This shouldn't be in your PR

**Review by humphd — CHANGES_REQUESTED** [review]

File: .github/workflows/algolia-scraper.yml (hunk 28 above)
Comment (ycechungAI) [reply .github/workflows/algolia-scraper.yml:2]: i see so same file i was like how do u call one from another

Comment (ycechungAI) [reply .github/workflows/algolia-scraper.yml:2]: image

Comment (ycechungAI) [reply .github/workflows/algolia-scraper.yml]: oh i think u combine run and env parts

File: CONTRIBUTING.md [hunk 31]
```diff
@@ -119,3 +125,22 @@ Tool Recommendation: Firefox Browser (easier to make screenshots)
 ```
 !["Alt test for accessibility"](/static/img/name-of-the-image.png)
 ```
+
+### Search
+
+We use Algolia to build our index. In order to build an index for search:
+
+1. sign up for an account on https://www.algolia.com/users/sign_in
```
Comment (humphd) [inline CONTRIBUTING.md:133]: @CameronGray1210 and I tried to follow these steps, but there's a bunch of missing info here.  Going through the Algolia site to create our index is more complicated than what is listed here.  I'd like to see this expanded.  For example, unless we upload a sample data set we can't move forward with the process.

**Review by humphd — CHANGES_REQUESTED** [review]

File: CONTRIBUTING.md (hunk 31 above)
Comment (ycechungAI) [reply CONTRIBUTING.md:128]: ya but if i copy the whole site info on here will be long as well, thats why i provided links
Since you only create the account once.

Comment (ycechungAI) [reply CONTRIBUTING.md:128]: i was under the impression if your site is live, it crawls the website automatically building the index,  I believe.  That is what the script does.

Comment (ycechungAI) [reply CONTRIBUTING.md:128]: I'm not sure if you need to upload the whole website onto Algo.. thats only if you have them build it for you

Comment (ycechungAI) [reply CONTRIBUTING.md:128]: we are building our own index

**Comment by ycechungAI** [comment]
rest of steps is from either website or using the UI which is hard to explain in words

**Comment by humphd** [comment]
@ycechungAI we met today to discuss possible directions for this work, and I think we're going to call this experiment "done" and pivot to using a pure client-side search index vs. server-based with Algolia.

I'm going to file a new issue on the approach, and assign to you.

Thanks for taking the lead on this.  Even though we aren't going this route, it was useful to explore it so we could eliminate it as our primary option.

**Comment by humphd** [comment]
Filed #125
